src/odemis/driver/smaract.py

```diff
@@ -3173,3 +2961,1175 @@ def SA_CTL_Move(self, handle, ch, pos_pm, _):
+        while state != end_state:
+            if tend:
+                t = int((tend - time.time()) * 1000)  # SA_SI_WaitForEvent accepts timeout in ms
+                if t < 0 or ev.error == SA_SI_ERROR_TIMEOUT:
+                    raise TimeoutError()
+                t = c_uint(t)
+            else:
+                t = SA_SI_TIMEOUT_INFINITE
+
+            # Don't check event type, it's not always accurate.
+            # Instead, only one event is active and all other events are disabled, so it is clear
+            # which event we are waiting for.
+            self.core.SA_SI_WaitForEvent(self._id, byref(ev), t)
+            state = ev.devEventParameter & 0xffff
+
+    @isasync
+    def reference(self, _=None):
+        """
+        This is not a "normal" referencing procedure since the hardware doesn't have any actuators. Instead,
+        it performs an internal adjustment routine which is required to get accurate position values.
+        """
+        f = self._createReferenceFuture()
+        self._executor.submitf(f, self._doReference, f)
+        return f
+
+    def _doReference(self, future):
+        """
+        Actually runs the referencing code.
+        future (Future): the future it handles
+        raise:
+            IOError: if referencing failed due to hardware
+            CancelledError if was cancelled
+        """
+        # The position polling thread can continue, it does not seem to interfere with the referencing procedure.
+        try:
+            with future._moving_lock:
+                if future._must_stop:
+                    future._was_stopped = True
+                # Reset reference so that if it fails, it states the axes are not referenced (anymore)
+                self.referenced._value = {a: False for a in self._channels.keys()}
+                logging.debug("Starting referencing.")
+                # Cannot go immediately to automatic adjustment --> first switch to manual adjustment
+                self.EnableEventNotification(SA_PS_AF_ADJUSTMENT_PROGRESS_EVENT)
+                self.SetProperty_i32(SA_PS_AF_ADJUSTMENT_STATE_PROP,
+                                     SA_PS_ADJUSTMENT_STATE_MANUAL_ADJUST)
+            self._wait_for_event(SA_PS_ADJUSTMENT_STATE_MANUAL_ADJUST)
+            with future._moving_lock:
+                if future._must_stop:
+                    future._was_stopped = True
+                # Activate working distance
+                self.SetProperty_i32(SA_PS_SYS_WORKING_DISTANCE_ACTIVATE_PROP,
+                                     SA_PS_WORKING_DISTANCE_SHRINK_MODE_LEFT_RIGHT)
+            # attribute is write-only and there is no corresponding event type (yet some waiting time is necessary)
+            # --> wait 1 s for command to be processed
+            startt = time.time()
+            while time.time() - startt < 1:
+                if future._must_stop:
+                    future._was_stopped = True
+                time.sleep(0.01)
```

Comment:
If it's just 1 second, just put `time.sleep(1) in the lock, and be done.

Reply:
I wanted to keep it as similar as possible to the waitForEvent functions: call the function in the moving_lock and wait outside (also allow canceling without having to wait a second)